### ISSUE TITLE
[ENH] Cleaning up Catch22

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/sktime#2254]
While refactoring, I cam across `Catch22`.

This is a bit of a mess (old legacy code structure) and should be refactored as follows, imo:
* `_transform` should return a `pd.DataFrame`, with columns being the names in `feature_names`
* `Catch22` should accept a parameter `features` such as used in various functions. This is a `str` or list of `str`, of codes in `feature_names`. If passed, the transformer will calculate only those features.
* `X_inner_mtype` should also accept `pd.DataFrame`, i.e., a single series (in which case no loop is run over instances)
* the docstring should describe all parameters properly.

Good first issue, I think - most of these changes should be straightforward, albeit a bit fiddly.

**Comment by fkiraly** [comment]
@benfulcher has pointed me to this installable version of catch22:
https://github.com/chlubba/catch22/
I suggested a discussion on the topic of the "final form" of catch22, and possibly on running some benchmarking experiments, say, against tsfresh and the like.

**Comment by benfulcher** [comment]
There is a group of people working with catch22, including some improvements in pull requests about to be incorporated. We may also in the next days move the repo to an organizational repo. It's in active development.
Note that catch22 is a subset of the larger hctsa (>7k features) set (which is independently being ported to python [here](https://github.com/fairscape/hctsa-py)).

**Comment by MatthewMiddlehurst** [comment]
Happy to hear there is a Python version of hctsa coming 🙂.

The chlubba/catch22 package is what we originally used in sktime as a soft dependency, but after some issues with differing results by OS and one other issue I can't remember exactly breaking tests, we decided to just re-implement it using numba.

If this has been resolves/people are willing to overlook it, i don't see why we can't reintroduce it as an alternative.

**Comment by benfulcher** [comment]
We now have a dedicated python repo: https://github.com/DynamicsAndNeuralSystems/pycatch22. May be best to develop any modifications you require onto this repo for consistency (or to add the OS dependency tests)?

**Comment by fkiraly** [comment]
> or to add the OS dependency tests

Any plans for CI/CD in the repo?

**Comment by benfulcher** [comment]
Being a static feature set, we don't anticipate changes, and are not proficient in implementing CI/CD, so this isn't on the cards.
Not sure why we need the transformer there…? Users can use pycatch22 to simply extract the features, and use a combination workflow with `sktime` to do any extras.

**Comment by fkiraly** [comment]
> Being a static feature set, we don't anticipate changes, and are not proficient in implementing CI/CD, so this isn't on the cards.

Yes, but even if it's static, underlying packages could still change their interfaces, or some other python verseion compatibility bug could occur which only solid CI/CD would catch.

Want some help? I'm sure if we try to set up some joint `sktime`/`catch22` event (I haven't been to Sydney yet! 😄 ), there would be people who would join and help with setting up CI/CD.

>Not sure why we need the transformer there…?

A transformer is just the equivalent of wrapping it in a form with a standardized connector, so you can use it easily as components of more complex algorithms, e.g., classifiers or forecasters.
Why you would want to have a pipeline is I think explained well here on sklearn:
https://scikit-learn.org/stable/modules/compose.html

In brief, it would help `catch-22` to be used as components of more complex algorithms and workflows easily, and allow people to construct classifiers, clusterers, forecasters, etc.

**Comment by BrandonKMLee** [comment]
Any progress on this vs `tsfresh`? It has gone kinda quiet since last year.

**Comment by fkiraly** [comment]
Hm, @BrandonKMLee, I think this is done by this PR: https://github.com/sktime/sktime/pull/3431

The issue was just accidentally left open.

I checked the code and:

* `Catch22` - a re-implementation from scratch - is now interface compliant, along the lines of this issue
* a direct interface to the python package by @benfulcher has been added, it is the estimator `Catch22Wrapper`. This is also interface compliant.

I'll leave this open for a few days of comments or until resolution, but it looks like this can be closed as "done".

**Comment by fkiraly** [comment]
Had another look - while the estimators are interface compliant now, these do not seem done yet:

* `_transform` should return a `pd.DataFrame`, with columns being the names in `feature_names`
* `X_inner_mtype` should also accept `pd.DataFrame`, i.e., a single series (in which case no loop is run over instances)
* the docstring should describe all parameters properly.
